crates/client/src/client/async_client.rs

```diff
@@ -570,21 +571,92 @@ pub trait ClientHandle: 'static + Clone + DnsHandle<Error = ProtoError> + Send {
     }
+
+    /// Download all records from a zone using a AXFR query.
+    ///
+    /// # Arguments
+    /// * `zone_origin` - the zone name to update, i.e. SOA name
+    fn zone_transfert(
+        &mut self,
+        zone_origin: Name,
+    ) -> ClientStreamXfr<<Self as DnsHandle>::Response> {
+        let message = update_message::zone_transfert(zone_origin);
+
+        ClientStreamXfr {
+            inner: self.send(message),
+            fused: false,
+        }
+    }
+}
+
+/// A stream result of a Client Request
+#[must_use = "stream do nothing unless polled"]
+pub struct ClientStreamingResponse<R>(pub(crate) R)
+where
+    R: Stream<Item = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static;
+
+impl<R> Stream for ClientStreamingResponse<R>
+where
+    R: Stream<Item = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static,
+{
+    type Item = Result<DnsResponse, ClientError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Poll::Ready(ready!(self.0.poll_next_unpin(cx)).map(|r| r.map_err(ClientError::from)))
+    }
 }
 
 /// A future result of a Client Request
-#[must_use = "futures do nothing unless polled"]
+#[must_use = "future do nothing unless polled"]
 pub struct ClientResponse<R>(pub(crate) R)
 where
-    R: Future<Output = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static;
+    R: Stream<Item = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static;
 
 impl<R> Future for ClientResponse<R>
 where
-    R: Future<Output = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static,
+    R: Stream<Item = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static,
 {
     type Output = Result<DnsResponse, ClientError>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        self.0.poll_unpin(cx).map_err(ClientError::from)
+        Poll::Ready(
+            match ready!(self.0.poll_next_unpin(cx)) {
+                Some(r) => r,
+                None => Err(ProtoError::from(ProtoErrorKind::Timeout)),
+            }
+            .map_err(ClientError::from),
+        )
+    }
+}
+
+/// A stream result of a Client Request
+#[must_use = "stream do nothing unless polled"]
+pub struct ClientStreamXfr<R>
+where
+    R: Stream<Item = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static,
+{
+    pub(crate) inner: R,
+    fused: bool,
+}
+
+impl<R> Stream for ClientStreamXfr<R>
+where
+    R: Stream<Item = Result<DnsResponse, ProtoError>> + Send + Unpin + 'static,
+{
+    type Item = Result<DnsResponse, ClientError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.fused {
+            return Poll::Ready(None);
+        }
+        Poll::Ready(ready!(self.inner.poll_next_unpin(cx)).map(|some| {
+            some.map(|ok| {
+                if let Some(answer) = ok.answers().last() {
+                    self.fused = matches!(answer.rr_type(), RecordType::SOA);
```

Comment:
This kinda lack validation. It will stop when a message end with an SOA, which is the required comportment, however it does not check for the starting SOA, nor does it assert there is no other SOA transmitted in the middle, which is illegal.